extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,1263 @@
+  }
+
+  @Override
+  public void stop(boolean stopGracefully)
+  {
+    Preconditions.checkState(started, "not started");
+
+    log.info("Beginning shutdown of KafkaSupervisor[%s]", dataSource);
+
+    try {
+      timer.cancel(); // stop recurring executions
+
+      Optional<TaskRunner> taskRunner = taskMaster.getTaskRunner();
+      if (taskRunner.isPresent()) {
+        taskRunner.get().unregisterListener(supervisorId);
+      }
+
+      // Stopping gracefully will synchronize the end offsets of the tasks and signal them to publish, and will block
+      // until the tasks have acknowledged or timed out. We want this behavior when we're explicitly shut down through
+      // the API, but if we shut down for other reasons (e.g. we lose leadership) we want to just stop and leave the
+      // tasks as they are.
+      if (stopGracefully) {
+        log.info("Stopping gracefully, signalling managed tasks to complete and publish");
+        synchronized (stopLock) {
+          notices.add(new ShutdownNotice());
+          while (!stopped) {
+            stopLock.wait();
+          }
+        }
+        log.info("Shutdown notice handled");
+      }
+
+      exec.shutdownNow();
+      consumer.close();
+      started = false;
+
+      log.info("KafkaSupervisor[%s] has stopped", dataSource);
+    }
+    catch (Exception e) {
+      log.makeAlert(e, "Exception stopping KafkaSupervisor[%s]", dataSource)
+         .emit();
+    }
+  }
+
+  @Override
+  public SupervisorReport getStatus()
+  {
+    return generateReport(true);
+  }
+
+  public void possiblyRegisterListener()
+  {
+    // getTaskRunner() sometimes fails if the task queue is still being initialized so retry later until we succeed
+
+    if (listenerRegistered) {
+      return;
+    }
+
+    Optional<TaskRunner> taskRunner = taskMaster.getTaskRunner();
+    if (taskRunner.isPresent()) {
+      taskRunner.get().registerListener(
+          new TaskRunnerListener()
+          {
+            @Override
+            public String getListenerId()
+            {
+              return supervisorId;
+            }
+
+            @Override
+            public void locationChanged(final String taskId, final TaskLocation newLocation)
+            {
+              // do nothing
+            }
+
+            @Override
+            public void statusChanged(String taskId, TaskStatus status)
+            {
+              notices.add(new RunNotice());
+            }
+          }, MoreExecutors.sameThreadExecutor()
+      );
+
+      listenerRegistered = true;
+    }
+  }
+
+  private interface Notice
+  {
+    void handle();
+  }
+
+  private class RunNotice implements Notice
+  {
+    @Override
+    public void handle()
+    {
+      long nowTime = System.currentTimeMillis();
+      if (nowTime - lastRunTime < MAX_RUN_FREQUENCY_MILLIS) {
+        return;
+      }
+      lastRunTime = nowTime;
+
+      runInternal();
```

Comment:
Hmm, I think ideally what I'd like to see at INFO is,

1) something on startup saying, hey, we started!
2) updates whenever something changes (when kafka partitions are added, when tasks are created or destroyed or paused or whatever, stuff like that)
3) periodic messages when `runInternal` runs. Maybe not every single run if it happens really often (like more than a couple times a minute is probably excessive) but enough such that I know from the logs that the supervisor is still waking up and not stuck. As a user I'd be looking for that if I expect it to be doing something that it's not doing – the first thing to check is whether it's wedged